### PR TITLE
Statistics Collection

### PR DESCRIPTION
**Pull request by malin1993ml**
This PR contains the function of collecting following statistics:
1. Number of txns committed/aborted by the system;
2. Average txn throughput;
3. Average txn latency;
4. Number of tuples read/inserted/updated/deleted in every database, table, and index.

Things TODO:
1. Write the aggregated stats to the internal catalog table.
2. Replace the thread_local stats context with a global concurrent hash map. (for compiling on Mac in the future).
3. Add query-level statistics.

### CODE REVIEW

File: script/oltpbenchmark/postgresql.conf [hunk 1]
```diff
@@ -637,3 +637,5 @@ peloton_checkpoint_mode invalid
 # Peloton log directory
 peloton_log_directory = 'peloton_log'
 
+# Stats collection mode: invalid / enable.
```
Comment (jarulraj) [inline script/oltpbenchmark/postgresql.conf]: Let's remove this file.

File: src/catalog/catalog.cpp [hunk 2]
```diff
@@ -157,8 +157,9 @@ Result Catalog::CreatePrimaryIndex(const std::string &database_name,
     key_schema->SetIndexedColumns(key_attrs);
 
     index_metadata = new index::IndexMetadata(
-        "customer_pkey", GetNextOid(), INDEX_TYPE_BWTREE,
-        INDEX_CONSTRAINT_TYPE_PRIMARY_KEY, schema, key_schema, key_attrs, true);
+        "customer_pkey", GetNextOid(), table->GetOid(), database->GetOid(),
+        INDEX_TYPE_BWTREE, INDEX_CONSTRAINT_TYPE_PRIMARY_KEY, schema,
+        key_schema, key_attrs, true);
```
Comment (jarulraj) [inline src/catalog/catalog.cpp]: use local variable "bool unique = true" and use unique as the argument, its more readable.

File: src/include/common/types.h [hunk 3]
```diff
@@ -770,6 +769,24 @@ enum CheckpointStatus {
...
 };
 
+// Statistics Collection Type
+// Disable or enable
+enum StatsType {
+  STATS_TYPE_INVALID = 0,
+  STATS_TYPE_ENABLE = 1,
+};
+
+enum MetricType {
```
Comment (jarulraj) [inline src/include/common/types.h]: Add some comments on types of metrics.

File: src/include/common/types.h [hunk 4]
```diff
@@ -770,6 +769,24 @@ enum CheckpointStatus {
...
+  TEMPORAL_METRIC = 0,
```
Comment (jarulraj) [inline src/include/common/types.h]: METRIC_TYPE_INVALID = 0

File: src/include/statistics/abstract_metric.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Abstract class for metrics
+ * A metric should be able to:
+ * (1) identify its type;
+ * (2) print itself (ToString);
+ * (3) reset itself;
+ * (4) aggregate itself with another source
+ *     of the same type.
+ */
+class AbstractMetric {
```
Comment (jarulraj) [inline src/include/statistics/abstract_metric.h]: Can you inherit from `Printable` ?

File: src/include/statistics/access_metric.h [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  inline CounterMetric &GetAccessCounter(size_t counter_type) {
+    return access_counters_[counter_type];
+  }
+
+  //===--------------------------------------------------------------------===//
+  // HELPER METHODS
+  //===--------------------------------------------------------------------===//
+
+  inline bool operator==(const AccessMetric &other) {
+    for (size_t i = 0; i < access_counters_.size(); ++i) {
```
Comment (jarulraj) [inline src/include/statistics/access_metric.h]: `i` -> `access_counter`

File: src/include/statistics/backend_stats_context.h [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+                         ->GetTableId();
+    oid_t database_id = catalog::Manager::GetInstance()
+                            .GetTileGroup(tile_group_id)
+                            ->GetDatabaseId();
+    auto table_metric = GetTableMetric(database_id, table_id);
+    assert(table_metric != nullptr);
+    table_metric->GetTableAccess().IncrementUpdates();
+  }
+
+  inline void IncrementTableDeletes(oid_t tile_group_id) {
```
Comment (jarulraj) [inline src/include/statistics/backend_stats_context.h]: Do we need to inline everything ?

File: src/include/statistics/backend_stats_context.h [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+  inline void IncrementTableInserts(oid_t tile_group_id) {
+    oid_t table_id = catalog::Manager::GetInstance()
+                         .GetTileGroup(tile_group_id)
+                         ->GetTableId();
+    oid_t database_id = catalog::Manager::GetInstance()
+                            .GetTileGroup(tile_group_id)
+                            ->GetDatabaseId();
+    auto table_metric = GetTableMetric(database_id, table_id);
+    assert(table_metric != nullptr);
```
Comment (jarulraj) [inline src/include/statistics/backend_stats_context.h]: Use `PL_ASSERT` instead of raw `assert`

File: src/include/statistics/latency_metric.h [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  //===--------------------------------------------------------------------===//
+  // MEMBERS
+  //===--------------------------------------------------------------------===//
+
+  // Circular buffer with capacity N that stores the <= N
+  // most recent latencies collected
+  boost::circular_buffer<double> latencies_;
+
+  // Timer for timing individual latencies
+  Timer<std::ratio<1, 1000>> timer_ms_;
```
Comment (jarulraj) [inline src/include/statistics/latency_metric.h]: Good, you are using our timer.

File: src/include/statistics/latency_metric.h [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  boost::circular_buffer<double> Copy();
+
+ private:
+  //===--------------------------------------------------------------------===//
+  // MEMBERS
+  //===--------------------------------------------------------------------===//
+
+  // Circular buffer with capacity N that stores the <= N
+  // most recent latencies collected
+  boost::circular_buffer<double> latencies_;
```
Comment (jarulraj) [inline src/include/statistics/latency_metric.h]: Add a container called "CircularBuffer" and then implement it using `boost::circular_buffer`.

File: src/include/statistics/stats_aggregator.h [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  //===--------------------------------------------------------------------===//
+
+  // Register the BackendStatsContext of a worker thread to global Stats
+  // Aggregator
+  inline void RegisterContext(std::thread::id id_,
+                              BackendStatsContext *context_) {
+    {
+      std::lock_guard<std::mutex> lock(stats_mutex_);
+
+      // FIXME: This is sort of hacky. Eventually we want to free the
```
Comment (jarulraj) [inline src/include/statistics/stats_aggregator.h]: Can you fix it now ?

File: test/include/statistics/stats_tests_util.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace peloton {
+namespace test {
+
+class StatsTestsUtil {
+ public:
+  static storage::Tuple PopulateTuple(const catalog::Schema *schema,
+                                      int first_col_val, int second_col_val,
+                                      int third_col_val, int fourth_col_val);
+};
```
Comment (jarulraj) [inline test/include/statistics/stats_tests_util.h]: Add namespace end.

File: test/statistics/stats_test.cpp [hunk 13]
```diff
@@ -0,0 +1,216 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// data_table_test.cpp
```
Comment (jarulraj) [inline test/statistics/stats_test.cpp]: Fix names.

File: test/statistics/stats_test.cpp [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+              .GetReads();
+  int64_t deletes = stats::BackendStatsContext::GetInstance()
+                        .GetTableMetric(database_id, table_id)
+                        ->GetTableAccess()
+                        .GetDeletes();
+  EXPECT_EQ(4, txn_commited);
+  EXPECT_EQ(9, reads);
+  EXPECT_EQ(1, deletes);
+}
+}
```
Comment (jarulraj) [inline test/statistics/stats_test.cpp]: Add namespace comments.

File: test/statistics/stats_test.cpp [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+        TransactionTestsUtil::MakeProjectInfoFromTuple(&tuple)};
+
+    // Insert
+    planner::InsertPlan node(data_table.get(), std::move(project_info));
+    std::unique_ptr<executor::ExecutorContext> context(
+        new executor::ExecutorContext(txn));
+    executor::InsertExecutor executor(&node, context.get());
+    executor.Execute();
+
+    /*
```
Comment (jarulraj) [inline test/statistics/stats_test.cpp]: Why is this commented out ?

File: test/statistics/stats_test.cpp [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+  // Create a table and wrap it in logical tiles
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+  std::unique_ptr<storage::DataTable> data_table(
+      ExecutorTestsUtil::CreateTable(tups_per_tile_group, true));
+
+  // Ensure that the tile group is as expected.
+  const catalog::Schema *schema = data_table->GetSchema();
+  assert(schema->GetColumnCount() == 4);
```
Comment (jarulraj) [inline test/statistics/stats_test.cpp]: PL_ASSERT

File: test/statistics/stats_test.cpp [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+//
+// Identification: tests/storage/basic_stats_test.cpp
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "common/harness.h"
+
+#include "statistics/stats_aggregator.h"
```
Comment (jarulraj) [inline test/statistics/stats_test.cpp]: Sort includes by directory name

File: test/statistics/stats_tests_util.cpp [hunk 18]
```diff
@@ -0,0 +1,50 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// transaction_tests_util.cpp
```
Comment (jarulraj) [inline test/statistics/stats_tests_util.cpp]: Fix names.

**Comment by malin1993ml** [comment]
All issues have been addressed.

**Pull request merged**